web/src/actions/activities.js

```diff
@@ -36,8 +42,112 @@ export const addActivity = () => (dispatch, getState) => {
   });
 };
 
-export const addActivityContractor = key =>
-  actionWithYears(ADD_ACTIVITY_CONTRACTOR, { key });
+export const updateActivity = (key, updates, isExpense = false) => dispatch => {
+  dispatch({
+    type: UPDATE_ACTIVITY,
+    key,
+    updates
+  });
+  if (isExpense) {
+    dispatch(updateBudget());
+  }
+};
+
+export const addActivityContractor = (key, { save = saveApd } = {}) => async (
+  dispatch,
+  getState
+) => {
+  const activity = getState().activities.byKey[key];
+  const oldIds = activity.contractorResources.map(c => c.id);
+
+  dispatch(actionWithYears(ADD_ACTIVITY_CONTRACTOR, { key }));
+
+  const newApd = await dispatch(save());
```

Comment:
Add the contractor, then save.  That way we let the dirty reducer do the work of making sure the contractor gets added to the data we save.